File: src/mkfs.hs
module Main where

import qualified Interpreter as I
import qualified AsyncFS
import FSLayout
import Disk
import System.Environment
import qualified Errno

cachesize :: Integer
cachesize = 100000

main :: IO ()
main = do
  args <- getArgs

  case args of
    [fn] -> do
      ds <- init_disk fn
      putStrLn $ "Initializing file system"
      res <- I.run ds $ AsyncFS._AFS__mkfs cachesize 4 1 256
      case res of
        Errno.Err _ -> error $ "mkfs failed"
        Errno.OK (_, fsxp) -> do
          putStrLn $ "Initialization OK, " ++ (show $ coq_FSXPMaxBlock fsxp) ++ " blocks"
          putStrLn $ "  === disk layout addresses ==="
          putStrLn $ "  superblock: 0"
          putStrLn $ "  data blocks start: 1"
          putStrLn $ "  inode blocks start: " ++ (show . (+1) . coq_IXStart . coq_FSXPInode $ fsxp)
          putStrLn $ "  log header: " ++ (show . coq_LogHeader . coq_FSXPLog $ fsxp)
          putStrLn $ "  log data start: " ++ (show . coq_LogData . coq_FSXPLog $ fsxp)

      stats <- close_disk ds
      print_stats stats
    _ -> do
      putStrLn $ "Usage: mkfs diskpath"

File: src/mlbin/mkfs.ml
module Errno = Fscq.Errno
module AsyncFS = Fscq.AsyncFS
module FSLayout = Fscq.FSLayout

let cachesize = Z.of_int 100_000

let do_mkfs fn =
  let ds = Fscq.Disk.init fn in
  Printf.printf "Initializing filesystem in %s\n" fn;
  let mkfs_prog = AsyncFS.AFS.mkfs cachesize (Z.of_int 4) (Z.of_int 1) (Z.of_int 256) in
  begin match Fscq.Interp.run ds mkfs_prog with
  | Errno.Err _ -> Printf.printf "mkfs failed"
  | Errno.OK (_, fsxp) ->
    Printf.printf "Initialization OK, %s blocks\n"
      (Z.to_string fsxp.FSLayout.coq_FSXPMaxBlock);
    Printf.printf "  === disk layout addresses ===\n";
    Printf.printf "  superblock: 0\n";
    Printf.printf "  data blocks start: 1\n";
    Printf.printf "  inode blocks start: %s\n"
      (Z.to_string (Z.succ FSLayout.(coq_IXStart fsxp.coq_FSXPInode)));
    Printf.printf "  log header: %s\n"
      (Z.format "%x" FSLayout.(coq_LogHeader fsxp.coq_FSXPLog));
    Printf.printf "  log data start: %s\n"
      (Z.format "%x" FSLayout.(coq_LogData fsxp.coq_FSXPLog))
  end;
  Fscq.Disk.close ds

let _ =
  match Sys.argv with
  | [| _ ; fn |] ->
    do_mkfs fn
  | _ ->
    Printf.printf "Usage: %s <disk path>\n" Sys.argv.(0)
